server/src/pages/shot/editor.js

```diff
@@ -155,13 +181,24 @@ exports.Editor = class Editor extends React.Component {
       return;
     }
 
+    this.history.pushDiff(this.imageCanvas, affectedArea);
+
     this.imageContext.globalCompositeOperation = (compositeOp || "source-over");
     this.imageContext.drawImage(incomingCanvas,
       affectedArea.left * this.devicePixelRatio,
       affectedArea.top * this.devicePixelRatio,
       affectedArea.width * this.devicePixelRatio,
       affectedArea.height * this.devicePixelRatio,
       affectedArea.left, affectedArea.top, affectedArea.width, affectedArea.height);
+
+    this.deriveButtonStates();
+  }
+
+  applyDiff(area, diffCanvas) {
+    this.imageContext.globalCompositeOperation = "source-over";
+    this.imageContext.drawImage(diffCanvas,
```

Comment:
I suspect this might be where the devicePixelRatio is being left out

Reply:
hmmm `scale()` should have been called on that drawing context at that point.

Also possibly related to #4453.